src/canvas_test.c

```diff
@@ -0,0 +1,158 @@
+#include "minunit.h"
+#include "canvas.h"
+
+static int cols = 2;
+static int rows = 3;
+static char load_str[] = "012345";
+
+static Canvas *c1, *c2;
+
+MU_TEST(test_canvas_new) {
+    c1 = canvas_new(rows, cols);
+    mu_assert(c1->num_cols == cols, "cols should be 2");
+    mu_assert(c1->num_rows == rows, "rows should be 3");
+    canvas_free(c1);
+}
+
+MU_TEST(test_canvas_load_str) {
+    c1 = canvas_new(rows, cols);
+    int numset = canvas_load_str(c1, load_str);
+    mu_assert(numset == sizeof(load_str) - 1, "numset should be 6");
+    mu_check(c1->rows[0][0] == '0');
+    mu_check(c1->rows[0][1] == '1');
+    mu_check(c1->rows[1][0] == '2');
+    mu_check(c1->rows[1][1] == '3');
+    mu_check(c1->rows[2][0] == '4');
+    mu_check(c1->rows[2][1] == '5');
+
+    canvas_free(c1);
+}
+
+MU_TEST(test_canvas_eq) {
+    // canvas with same content
+    c1 = canvas_new(rows, cols);
+    c2 = canvas_new(rows, cols);
+
+    // canvases of different sizes
+    Canvas *c3 = canvas_new(cols - 1, rows);
+    Canvas *c4 = canvas_new(cols, rows - 1);
+
+    // init all canvases with same str
+    canvas_load_str(c1, load_str);
+    canvas_load_str(c2, load_str);
+    canvas_load_str(c3, load_str);
+    canvas_load_str(c4, load_str);
+
+    // same canvases should be the same
+    mu_check(canvas_eq(c1, c2));
+    mu_check(canvas_eq(c2, c1));
+
+    // different canvases should be different
+    mu_check(!canvas_eq(c1, c3));
+    mu_check(!canvas_eq(c1, c4));
+    mu_check(!canvas_eq(c3, c4));
+
+    // change one of the same canvases
+    c1->rows[2][1] = 'A';
+    mu_check(!canvas_eq(c1, c2));
+
+    // change the other to match
+    c2->rows[2][1] = 'A';
+    mu_check(canvas_eq(c1, c2));
+
+    canvas_free(c1);
+    canvas_free(c2);
+    canvas_free(c3);
+    canvas_free(c4);
+}
+
+// Tests to make sure the setup/teardown and eq functions will work
+MU_TEST_SUITE(canvas_bootstrapping) {
+    MU_RUN_TEST(test_canvas_new);
+    MU_RUN_TEST(test_canvas_load_str);
+    MU_RUN_TEST(test_canvas_eq);
+}
+
```

Comment:
```suggestion

// Begin canvas_main tests
```
This jump confused me a little, I was thinking, "Why suddenly define some functions?"